### PR TITLE
Web Worker Decrypt

### PR DESCRIPTION
**Pull request by aitrean**
Closes https://github.com/MyEtherWallet/MyEtherWallet/issues/423, https://github.com/MyEtherWallet/MyEtherWallet/issues/219

This PR includes three solutions. 
1. Attempt an empty password every time a keystore is uploaded.
2. Delegate scrypt decryption (ie ethereumjs-wallet.fromV3) to its own web worker and interface with it through an async typescript function that gets handled in the wallet saga. This keeps the UI unblocked when scrypt takes a long time to decrypt.
3. Add logic to show a spinner x number of milliseconds after file upload so the user will understand when a wallet is being decrypted.

### CODE REVIEW

File: common/components/WalletDecrypt/Keystore.tsx [hunk 1]
```diff
@@ -21,12 +22,14 @@ function isPassRequired(file: string): boolean {
...
   public props: {
     value: KeystoreValue;
+    isWalletLoading: boolean;
     onChange(value: KeystoreValue): void;
     onUnlock(): void;
   };
 
   public render() {
     const { file, password } = this.props.value;
+    const isWalletLoading = this.props.isWalletLoading;
```
Comment (dternyak) [inline common/components/WalletDecrypt/Keystore.tsx]: Why not de-structure?

**Comment by dternyak** [comment]
Impressive work! Going to leave it open a bit more and solicit feedback from @HenryNguyen5 and @wbobeirne who participated in design.

**Comment by dternyak** [comment]
Here's an encrypted Keystore with no password for testing purposes: 
[UTC--2017-12-30T16-45-58.340Z--80965bfdbb6c44c8ed652add058bf372e2b9dc86.zip](https://github.com/MyEtherWallet/MyEtherWallet/files/1594575/UTC--2017-12-30T16-45-58.340Z--80965bfdbb6c44c8ed652add058bf372e2b9dc86.zip)

File: common/libs/web-workers/workers/scrypt-worker.worker.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { fromV3, IFullWallet } from 'ethereumjs-wallet';
+
+declare var postMessage: any;
+interface DecryptionParameters {
+  keystore: string;
+  password: string;
+  nonStrict: boolean;
+}
+
+onmessage = event => {
```
Comment (dternyak) [inline common/libs/web-workers/workers/scrypt-worker.worker.ts]: event could use typing

File: package.json [hunk 3]
```diff
@@ -110,7 +110,8 @@
     "webpack": "3.10.0",
     "webpack-dev-middleware": "2.0.3",
     "webpack-hot-middleware": "2.21.0",
-    "webpack-sources": "1.0.1"
+    "webpack-sources": "1.0.1",
+    "worker-loader": "^1.1.0"
```
Comment (wbobeirne) [inline package.json]: Lock down version # (We try to avoid `^` versions to keep builds consistent.)

File: common/libs/web-workers/scrypt-wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { IFullWallet, fromPrivateKey } from 'ethereumjs-wallet';
+import { toBuffer } from 'ethereumjs-util';
+const Worker = require('worker-loader!./workers/scrypt-worker.worker.ts');
```
Comment (wbobeirne) [inline common/libs/web-workers/scrypt-wrapper.ts]: We should setup typing for this, as outlined here: https://github.com/webpack-contrib/worker-loader#integrating-with-typescript. Right now it's an implicit `any`.

File: common/actions/wallet/actionCreators.ts [hunk 5]
```diff
@@ -43,6 +43,13 @@ export function setWallet(value: IWallet): types.SetWalletAction {
   };
 }
 
+export function setWalletLoading(loadingStatus: boolean): types.SetWalletLoadingAction {
```
Comment (wbobeirne) [inline common/actions/wallet/actionCreators.ts]: Existing nomenclature would be `setWalletPending`

File: common/components/WalletDecrypt/Keystore.tsx [hunk 6]
```diff
@@ -57,6 +60,7 @@ export default class KeystoreDecrypt extends Component {
                 type="password"
               />
             </div>
+            {isWalletLoading ? <Spinner /> : ''}
```
Comment (wbobeirne) [inline common/components/WalletDecrypt/Keystore.tsx]: Minor, but you can simplify these boolean checks to `{isWalletLoading && <Spinner />}`

File: common/libs/web-workers/workers/scrypt-worker.worker.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { fromV3, IFullWallet } from 'ethereumjs-wallet';
+
+declare var postMessage: any;
```
Comment (wbobeirne) [inline common/libs/web-workers/workers/scrypt-worker.worker.ts]: From https://github.com/webpack-contrib/worker-loader#integrating-with-typescript:

```
const ctx: Worker = self as any;
ctx.postMessage({ foo: "foo" });
```
for typing

File: common/sagas/wallet/wallet.ts [hunk 8]
```diff
@@ -141,18 +146,43 @@ export function* unlockPrivateKey(action: UnlockPrivateKeyAction): SagaIterator
   yield put(setWallet(wallet));
 }
 
+export function* startLoadingSpinner(): SagaIterator {
+  yield call(delay, 1000);
```
Comment (wbobeirne) [inline common/sagas/wallet/wallet.ts]: I'd probably shorten this to 300-500, a full second before having the site respond is a little long.

File: common/libs/web-workers/scrypt-wrapper.ts (hunk 4 above)
Comment (wbobeirne) [reply common/libs/web-workers/scrypt-wrapper.ts]: Still typed as any, but
```
import Worker from 'worker-loader!./workers/scrypt-worker.worker.ts';
```
seems to fix it!
